src/bun.js/api/server.zig

```diff
@@ -5796,6 +5796,177 @@ pub const ServerWebSocket = struct {
+        for (this.value.pending.raw_plugins, 0..) |raw_plugin, i| {
+            bunstring_array[i] = bun.String.init(raw_plugin);
+        }
+        const plugins = bun.String.toJSArray(globalThis, bunstring_array);
+        const bunfig_folder_bunstr = bun.String.createUTF8ForJS(globalThis, bunfig_folder);
+
+        globalThis.bunVM().eventLoop().enter();
+        const result = JSBundlerPlugin__loadAndResolvePluginsForServe(plugin, plugins, bunfig_folder_bunstr);
+        globalThis.bunVM().eventLoop().exit();
+        // handle the case where js synchronously throws an error
+        if (globalThis.tryTakeException()) |e| {
+            handleOnReject(this, globalThis, e);
+            return;
+        }
+
+        if (!result.isEmptyOrUndefinedOrNull()) {
+            // handle the case where js returns a promise
+            if (result.asAnyPromise()) |promise| {
+                switch (promise.status(globalThis.vm())) {
+                    // promise not fulfilled yet
+                    .pending => {
+                        this.ref();
+                        this.value.pending.promise.strong.set(globalThis, promise.asValue(globalThis));
+                        promise.asValue(globalThis).then(globalThis, this, onResolveImpl, onRejectImpl);
+                        return;
+                    },
+                    .fulfilled => {
+                        handleOnResolve(this);
+                        return;
+                    },
+                    .rejected => {
+                        // const value = promise.asValue(globalThis);
+                        const value = promise.result(globalThis.vm());
+                        handleOnReject(this, globalThis, value);
+                        return;
+                    },
+                }
+            }
+
+            if (result.toError()) |e| {
+                handleOnReject(this, globalThis, e);
+            } else {
+                handleOnResolve(this);
+            }
+        }
+    }
+
+    pub fn deinit(this: *ServePlugins) void {
+        if (this.value == .err) {
+            bun.default_allocator.free(this.value.err);
+        }
+        ServePlugins.destroy(this);
+    }
+
+    pub const onResolve = JSC.toJSHostFunction(onResolveImpl);
+    pub const onReject = JSC.toJSHostFunction(onRejectImpl);
+
+    pub fn onResolveImpl(_: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSValue {
+        ctxLog("onResolve", .{});
+
+        const arguments = callframe.arguments_old(2);
+        var plugins = arguments.ptr[1].asPromisePtr(ServePlugins);
+        defer plugins.deref();
+        const plugins_result = arguments.ptr[0];
+        plugins_result.ensureStillAlive();
+
+        handleOnResolve(plugins);
+
+        return JSValue.jsUndefined();
+    }
+
+    pub fn handleOnResolve(this: *ServePlugins) void {
+        this.value.pending.promise.deinit();
+        for (this.value.pending.pending_bundled_routes.items) |route| {
```

Comment:
can we swap the values here to make this durable to any kind of recursion


```suggestion
        var pending_bundled_routes = this.value.pending.pending_bundled_routes;
        defer pending_bundled_routes.deinit();
        this.value.pending.pending_bundled_routes = .{};
        for (pending_bundled_routes.items) |route| {
```